### PR TITLE
amqp - kit new logic and api

### PR DESCRIPTION
**Pull request by akolomietc**
update vault env variable

### CODE REVIEW

File: amqp-kit/channel.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+		return conn, nil
+	default:
+		c := &channel{}
+		c.c, c.err = p.c.Channel()
+		return c, c.err
+	}
+}
+
+func (p *pool) put(c *channel) {
+	if c.err == nil {
```
Comment (s0rg) [inline amqp-kit/channel.go]: It better to invert this logic (and use early return) to reduce the cyclomatic complexity

File: amqp-kit/channel.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+		case p.ch <- c:
+		default:
+			c.close()
+		}
+	}
+}
+
+func (p *pool) empty() {
+	var c *channel
+	for {
```
Comment (s0rg) [inline amqp-kit/channel.go:56]: I think, its better to use `range` here instead of `for + select` pair

File: amqp-kit/client.go [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+func (c *Client) onCloseWithErr(conn *connection, err error) {
+
+	log.Warnf("AMQP: connection closed, err %v", err)
+
+	for {
+		select {
+		case <-c.stopClientChan:
+			return
+		default:
+			<-time.After(2 * time.Second)
```
Comment (s0rg) [inline amqp-kit/client.go]: Move it constants, please

File: amqp-kit/client.go [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+			s.Workers = 1
+		}
+
+		subscribers[si.Queue] = &s
+	}
+
+	log.Info(`subscribers checked`)
+
+	for _, sub := range subscribers {
+		for i := 1; i <= sub.Workers; i++ {
```
Comment (s0rg) [inline amqp-kit/client.go]: Why? You already patched value of sub.Workers (0 -> 1)

**Review by s0rg — APPROVED** [review]

File: amqp-kit/channel.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	c  *amqp.Connection
+}
+
+type channel struct {
+	c   *amqp.Channel
+	err error
+}
+
+func (c *channel) close() {
+	if c != nil && c.c != nil {
```
Comment (avdva) [inline amqp-kit/channel.go]: what is the point of check if `this` is nil?

File: amqp-kit/channel.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+	}
+
+	select {
+	case p.ch <- c:
+	default:
+		c.close()
+	}
+}
+
+func (p *pool) empty() {
```
Comment (avdva) [inline amqp-kit/channel.go]: its much, much better to call it `clear` instead of `empty`

File: amqp-kit/client.go [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+	"time"
+
+	"github.com/go-kit/kit/endpoint"
+	log "github.com/sirupsen/logrus"
+	"github.com/streadway/amqp"
+)
+
+const defaultReconnectAfterDuration = 2 * time.Second
+
+type Publisher interface {
```
Comment (avdva) [inline amqp-kit/client.go:19]: no comments for public entities?
or its ok in your team's style-guide?

File: amqp-kit/client.go [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+	u := url.URL{Scheme: "amqp", User: url.UserPassword(c.User, c.Password), Host: c.Address, Path: "/" + c.VirtualHost}
+	return u.String()
+}
+
+func (si *SubscribeInfo) keyName() string {
+	return strings.Replace(si.Queue, "_", ".", -1)
+}
+
+func (c *Client) init() error {
+	err := c.reconnect()
```
Comment (avdva) [inline amqp-kit/client.go]: `return c.reconnect` ?
you can remove this func at all.

File: amqp-kit/client.go [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+
+	log.Warnf("AMQP: connection closed, err %v", err)
+
+	for {
+		select {
+		case <-c.stopClientChan:
+			return
+		default:
+			afterDuration := c.config.ReconnectAfterDuration
+			if afterDuration.Seconds() == 0 {
```
Comment (avdva) [inline amqp-kit/client.go]: `afterDuration == 0` ?

File: amqp-kit/client.go [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+				afterDuration = defaultReconnectAfterDuration
+			}
+			<-time.After(afterDuration)
```
Comment (avdva) [inline amqp-kit/client.go]: time.Sleep()

File: amqp-kit/client.go [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+func (c *Client) onCloseWithErr(conn *connection, err error) {
+
+	log.Warnf("AMQP: connection closed, err %v", err)
+
+	for {
+		select {
+		case <-c.stopClientChan:
+			return
+		default:
+			afterDuration := c.config.ReconnectAfterDuration
```
Comment (avdva) [inline amqp-kit/client.go]: the entire `afterDuration` calculation can be moved out of the `for` loop

File: amqp-kit/client.go [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+	msgs, err := ch.c.Consume(si.Queue, si.Name, false, false, false, false, nil)
+	if err != nil {
+		return fmt.Errorf(" Channel consume err: %s", err.Error())
+	}
+	fun := NewSubscriber(si.E, si.Dec, si.Enc, si.O...).ServeDelivery(ch.c)
+
+	var d amqp.Delivery
+	var ok bool
+
+	for {
```
Comment (avdva) [inline amqp-kit/client.go]: why not `for range` loop?

File: amqp-kit/client.go [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+	_, err = ch.QueueDeclare(queue, true, false, false, false, nil)
+	if err != nil {
+		return err
+	}
+
+	if err := ch.Qos(qos, 0, false); err != nil {
+		return err
+	}
+
+	if err := ch.QueueBind(queue, key, exchange, false, nil); err != nil {
```
Comment (avdva) [inline amqp-kit/client.go]: return ch.QueueBind(queue, key, exchange, false, nil)

File: amqp-kit/client.go [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+		return err
+	}
+
+	return nil
+}
+
+// Publish publishing some message to given exchange with key and correlationID
+func (c *Client) Publish(exchange, key, corID string, body []byte) (err error) {
```
Comment (avdva) [inline amqp-kit/client.go]: why named return variable, if you don't use it?

File: amqp-kit/client.go [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+		ContentType:   "application/json",
+		CorrelationId: corID,
+		Body:          body,
+		DeliveryMode:  amqp.Persistent,
+	}
+
+	return c.send(exchange, key, &pub)
+}
+
+func (c *Client) send(exchange, key string, pub *amqp.Publishing) (err error) {
```
Comment (avdva) [inline amqp-kit/client.go]: why named `err`?

File: amqp-kit/conn.go [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+	amqpConn     *amqp.Connection
+	pool         *pool
+	poolLock     sync.RWMutex
+	closeHandler connectionCloseHandler
+}
+
+func newConnection(config *Config, closeHandler connectionCloseHandler) *connection {
+	conn := &connection{
+		config:       config,
+		poolLock:     sync.RWMutex{},
```
Comment (avdva) [inline amqp-kit/conn.go]: useless line

**Pull request merged**